src/poll.c

```diff
@@ -0,0 +1,372 @@
+
+	poll_thread = pthread_self();
+
+	// Recieve signal USR2 and add a signal handler.
+	{
+		sigset_t sigset;
+		sigemptyset(&sigset);
+		sigaddset(&sigset, SIGUSR2);
+
+		if(pthread_sigmask(SIG_UNBLOCK, &sigset, NULL) == -1) {
+			error("Could not unblock USR2 for the polling thread");
+		}
+
+		struct sigaction sa;
+		sigemptyset(&sa.sa_mask);
+		sa.sa_handler = poll_handler;
+		sa.sa_flags = 0;
+		if(sigaction(SIGUSR2, &sa, NULL) == -1) {
+			error("Failed to change signal handler for SIGHUP");
+		}
+	}
+
+	fd_set readfds, writefds, exceptfds;
+	struct poll_file *p;
+	int nfds;
+
+	// Infinite Poll
+	while(1) {
+		FD_ZERO(&readfds);
+		FD_ZERO(&writefds);
+		FD_ZERO(&exceptfds);
+		nfds = 0;
+
+		//-------------------------------------------------------------------------
+		// Initialize select
+		poll_file_list_lock();
+		for(p = poll_file_head; p ; p = p->next) {
+			switch(p->type) {
+				case POLLIN:
+					FD_SET(p->fd, &readfds);
+					break;
+				case POLLOUT:
+					FD_SET(p->fd, &writefds);
+					break;
+				case POLLERR:
+					FD_SET(p->fd, &exceptfds);
+					break;
+				default: continue;
+			}
+			if(p->fd > nfds)
+				nfds = p->fd;
+		}
+		poll_file_list_unlock();
+		//-------------------------------------------------------------------------
+
+		select(nfds + 1, &readfds, &writefds, &exceptfds, NULL);
+
+		//-------------------------------------------------------------------------
+		// Handle polled files
+		poll_file_list_lock();
+		for(p = poll_file_head; p ; p = p->next) {
+			switch(p->type) {
+				case POLLIN:
+					if(FD_ISSET(p->fd, &readfds)) {
+						poll_time_update_nolock(&p->tv);
+					}
+					break;
+				case POLLOUT:
+					if(FD_ISSET(p->fd, &writefds)) {
+						poll_time_update_nolock(&p->tv);
+					}
+					break;
+				case POLLERR:
+					if(FD_ISSET(p->fd, &exceptfds)) {
+						poll_time_update_nolock(&p->tv);
+					}
+					break;
+			}
+		}
+		poll_file_list_unlock();
+		//-------------------------------------------------------------------------
+	}
+}
+
+// ---------------------------------------------------------------------------
+// API - basic methods
+void *poll_file_register(char *path, int type) {
+	struct poll_file *p;
+
+	// Search if we already poll this file.
+	poll_file_list_lock();
+	p = poll_file_list_search_nolock(path, type);
+	poll_file_list_unlock();
+
+	// If not found, init and add poll_file.
+	if(!p) {
+		p = poll_file_init(path, type);
+		poll_file_list_lock();
+		poll_file_list_prepend_nolock(p);
+		poll_file_list_unlock();
+	}
+
+	// Add a new checker
+	poll_file_list_lock();
+	struct poll_check *retv = poll_check_init_nolock(p);
+	poll_file_list_unlock();
+
+	// Notify poll thread to start watching the new file
+	poll_interrupt();
+
+	return retv;
+}
+
+unsigned long long poll_occured(void *poll_descriptor) {
+	struct poll_check *p_check = poll_descriptor;
+	unsigned long long retv;
+
+	poll_file_list_lock();
+	retv = poll_time_difference_nolock(&p_check->poll_file->tv, &p_check->tv);
```

Comment:
"dirty" is something bad. T
There is no guarantee that the write `gettimeofday` is atomic. Even there is no guarantee that a write to the underlying data types of timeval are atomic.

Having a copy  will not solve the problem. We always will have to write to the copy in another thread than reading it. If we want to do it in a save way we need to synchronize this calls.

When thinking about removing the lock here we should think what could happen. We could have a dirty read. This means we read data we did not write. This can lead to a wrong return value of `poll_occured`. This can only happen when the file got polled and we check the polling while updating the timeval to check. There can be two errors:

1. We miss the poll in this call. The next call will get it.
2. We get a wrong return value. For now this seems ok to.

Because of the little probability this might happen and the little impact if it happens I would remove the lock here.

I do not know how much performance locking costs....